auth/localkeys.go

```diff
@@ -0,0 +1,288 @@
+		if err != nil {
+			return err
+		}
+		err = DumpPEMKeyPairToFile(filename, pub, priv)
+		if err != nil {
+			return err
+		}
+
+	} else if err != nil {
+		return err
+	}
+
+	publicKey, privateKey, err := LoadKeyPairFromFile(filename)
+	if err != nil {
+		return err
+	}
+
+	masterKeys = MasterKeys{publicKey, privateKey}
+
+	return nil
+}
+
+// DumpPEMKeyPairToFile dumps PEM-encoded public and private keys to a single file
+func DumpPEMKeyPairToFile(filename string, public, private []byte) error {
+	data, err := DumpRSAPEMKeyPair(public, private)
+	if err != nil {
+		return err
+	}
+	return ioutil.WriteFile(filename, data, 0666)
+}
+
+// LoadPEMKeyPair loads a private/public key pair from a reader over PEM-encoded data.
+//  The private key is first, the public key is second.
+func LoadKeyPairFromFile(filename string) (public crypto.PublicKey, private crypto.PrivateKey, err error) {
+	data, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, nil, err
+	}
+	return LoadRSAKeyPairPackage(data)
+}
+
+// LoadDelegateKeysFromPEM loads the local delegate keys (master public, delegate private)
+//  from PEM data passed in directly
+//  Useful mostly for writing tests
+func LoadDelegateKeysFromPEM(public, private []byte) error {
+	pub, priv, err := LoadRSAKeyPair(public, private)
+	if err != nil {
+		return err
+	}
+	updateDelegateKeys(pub, priv)
+	return nil
+}
+
+// LoadMasterKeysFromPEM loads the local master keys from PEM data passed in directly
+//  Useful mostly for writing tests
+func LoadMasterKeysFromPEM(public, private []byte) error {
+	pub, priv, err := LoadRSAKeyPair(public, private)
+	if err != nil {
+		return err
+	}
+
+	masterKeys = MasterKeys{pub, priv}
+	return nil
+}
+
+// ClearKeys wipes the current state
+func ClearKeys() {
+	masterKeys = MasterKeys{}
+	updateDelegateKeys(nil, nil)
+}
+
+// WaitForDelegateKeys blocks until delegate keys are defined.
+func WaitForDelegateKeys() {
+	dKeyCond.L.Lock()
+	defer dKeyCond.L.Unlock()
+	for delegateKeys.localPrivate == nil || delegateKeys.masterPublic == nil {
+		dKeyCond.Wait()
+	}
+}
+
+// WatchDelegateKeyFile watches the delegate key file on the filesystem and
+// updates the internal delegate keys when changes are detected.
+func WatchDelegateKeyFile(filename string, cancel chan interface{}) error {
+	filename = filepath.Clean(filename)
+
+	log := log.WithFields(logrus.Fields{
+		"keyfile": filename,
+	})
+
+	loadKeys := func() {
+		if err := LoadDelegateKeysFromFile(filename); err != nil {
+			log.WithError(err).Warn("Unable to load delegate keys from file. Continuing to watch for changes")
+		} else {
+			log.Info("Loaded delegate keys from file")
+		}
+	}
+
+	// Try an initial load without any file changes
+	loadKeys()
+
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer w.Close()
+	cleanDir, _ := filepath.Split(filename)
+	ops := fsnotify.Write | fsnotify.Create
+	if err := w.Add(cleanDir); err != nil {
+		return err
+	}
+	for {
```

Comment:
It did when there, but this function has since been extracted into a utility and rewritten. That section is now in a goroutine, because that makes it callable easier.